Artifacts/windows-domain-join-new/artifact.ps1

```diff
@@ -0,0 +1,192 @@
+        $Username = 'WinNT://', "$DomainResolved", '/', $ADResolved -join ''
+    }
+    else
+    {
+        $ADResolved = ($Username -split '\\')[1]
+        $DomainResolved = ($Username -split '\\')[0]
+        $Username = 'WinNT://', $DomainResolved, '/', $ADResolved -join ''
+    }
+    return $Username
+}
+
+##############################
+# Function to add an AD user to a local group
+function Add-UserToLocalGroup ([Parameter(Mandatory = $true)] [string] $Username, [Parameter(Mandatory = $true)] [string] $GroupName)
+{
+    Write-Output "Attempting to add $DomainJoinUsername to the administrators group..."
+    $adUser = Get-AdUsernameString $Username
+    $group = [ADSI]("WinNT://$env:COMPUTERNAME/$GroupName, group")
+
+    if ($group.IsMember($adUser))
+    {
+        Write-Output "Result: $Username already belongs to the $GroupName"
+        return
+    }
+    $group.Add($adUser)
+    
+    if ($group.IsMember($adUser))
+    {
+        Write-Output "Result: $Username successfully added to $GroupName"
+    }
+    else
+    {
+        Write-Error "Result: failed to add $username to $GroupName"
+    }
+}
+
+##############################
+# Join computer to the domain
+function Add-VmToDomain ([Parameter(Mandatory = $true)] [string] $VmName, 
+    [Parameter(Mandatory = $true)] [string] $DomainName, 
+    [Parameter(Mandatory = $true)] [string] $JoinUser, 
+    [Parameter(Mandatory = $true)] [securestring] $JoinPassword, 
+    [Parameter(Mandatory = $true)] [string] $OU)
+{
+    Write-Output "Attempting to join the domain..."
+
+    if ((Get-WmiObject Win32_ComputerSystem).Domain -eq $DomainToJoin)
+    {
+        Write-Output "Computer is already joined to $DomainToJoin"
+    }
+    else
+    {
+        $credential = New-Object System.Management.Automation.PSCredential($JoinUser, $securePass)
+    
+        [Microsoft.PowerShell.Commands.ComputerChangeInfo]$computerChangeInfo = Add-Computer -ComputerName $VmName -DomainName $DomainName -Credential $credential -OUPath $OU -Force -PassThru
+        if ($computerChangeInfo.HasSucceeded)
+        {
+            Write-Output "Result: Successfully joined the $DomaintoJoin domain"
+        }
+        else
+        {
+            Write-Error "Result: Failed to join $env:COMPUTERNAME to $DomaintoJoin domain"
+        }
+    }
+}
+
+##############################
+# Main function
+
+if ($PSVersionTable.PSVersion.Major -lt 3)
+{
+    Write-Error "The current version of PowerShell is $($PSVersionTable.PSVersion.Major). Prior to running this artifact, ensure you have PowerShell 3 or higher installed."
+}
+
+else
+{
+    $adminGroupName = Get-LocalAdminGroupName
+    $securePass = ConvertTo-SecureString $DomainAdminPassword -AsPlainText -Force
+    Add-VmToDomain -VmName $env:COMPUTERNAME -DomainName $DomainToJoin -JoinUser $DomainAdminUsername -JoinPassword $securePass -OU $OUPath
+    Add-UserToLocalGroup -Username $DomainJoinUsername -GroupName $adminGroupName
```

Comment:
Let's not mix 2 artifacts' functionality into 1. I won't review the part dealing with adding the user to the local Administrators group, as all the comments already done in that artifact apply. Let the caller chain these 2 artifacts, or make a single artifact and get rid of the other one. No real advantage duplicating the code here, I don't think.